rfcs/47-better-error-handing.md

```diff
@@ -0,0 +1,251 @@
+
+### Alternative 1: Single Top-level Error and Multiple Middle-level Errors
+
+We can provide a single top-level `Error` type as below, and old top-level errors are turned into middle-level errors.
+
+```go
+type Error struct {
+	Op string
+	Err error
+}
+```
+
+And we allow zero or more error `struct`s to be nested.
+
+#### Pros
+
+- The user can get `Op` by `As(Error)`, instead of figuring out which top-level error it is. (But `Op` may be rare to be need elsewhere than in an error message.)
+- What's more, we can add more common fields like HTTP status code in the top-level error.
+- Multiple middle-level errors are more expressive.
+
+#### Cons 
+
+- Introduce break change.
+- Too many levels may be too complex both for implementors and users.
+
+### Alternative 2: Wrapping all SDK errors
+
+Wrapping SDK errors partially is bad. Then besides never wrapping them, we can also always wrap them. 
+
+#### Pros 
+
+- Do not lose error information.
+- The user can handle original SDK errors transparently with `As` & `Is`.
+
+#### Cons
+
+- Introduce break change: former bottom-level sentinel errors like `ErrObjectNotExist` should wrap an original error now.
+
+## Rationale
+
+First, it is necessary to state more clearly the current practice and turn it into a specification.
+
+The current situation has almost reached the goal of letting users know "where" and "why" an error happens:
+- The top-level errors' `Op` and their names tell users "where".
+- The context fields and the wrapped error can provide rich error message, telling users "why".
+
+But we can do more: let users handle error gracefully. The following can be done:
+- Help users quickly identify errors: 
+  - Ad-hoc string error can not be identified, so we should ban them.
+  - If we have too many levels, the users may use too general or too specific errors. We should provide proper granularity. The proposal's granularity can be described as:
+    - Top-level errors decide which component the errors belong to.
+    - Wrapped errors are specific. There isn't a more specific error type.
+    - If some wrapped errors can be grouped together, their `Unwrap` will return a sentinel error representing the category.
+- Provide a unified user experience: define an abstract layer of errors for the users, free them of the tedium of handling similar errors from multiple SDKs.
+
+## Compatibility
+
+The following changes will be made:
+- Extract some ad-hoc string errors into defined sentinel errors. (forward compatible)
+- Turn SDK errors into `fmt.Errorf("%v", err)`. (not forward compatible, but doesn't violate our promise)
+- Some error `struct`s' `Err` field will be removed. (not forward compatible)
+
+## Implementation
+
+Most of the work would be done by the author of this proposal.
```

Comment:
Please detail the works here and convert them into an issue after this proposal has been approved.